### DISCUSSION TITLE
IObservableArray annotation

### DISCUSSION THREAD

**Discussion by edshav** [mobxjs/mobx#2581] [Q&A] [answered]
Is there a better way to annotate `IObservableArray` to use its unique methods like `remove`?
```
import { IObservableArray, makeAutoObservable } from "mobx";
import { Item } from "./ItemModel";

class ItemList {
 constructor(public items: Item[] = []) {
    makeAutoObservable(this);
  }  

addItem = (value: string) => {
    this.items.push(new Item(value, this));
  };

removeItem = (itemToRemove: Item) => {
    (this.items as IObservableArray<Item>).remove(itemToRemove);
  };
}
```

**Comment by danielkcz** [comment]
By better, you mean what exactly? Is this not working for you?

**Reply by edshav** [reply]
This is working for me, but is there a way to annotate everything at once?

**Reply by edshav** [reply]
Is there any syntax that allows typescript to set types by inferring?

**Reply by danielkcz** [reply]
Well, you don't have `items` defined as a class property. In theory, you can declare the type on it instead of in each method. Inference won't probably work here.

**Comment by mweststrate** [comment, marked as answer]
Public modifier declares it as class prop implicitly. I'd declare it as
IObservableArray<Item> instead explicitly and cast the initial assignment

Op zo 1 nov. 2020 20:53 schreef Daniel K. <notifications@github.com>:

**Reply by edshav** [reply]
How can I cast the initial assignment as `IObservableArray<Item>` ? 
![image](https://user-images.githubusercontent.com/31659942/97863618-f0681500-1d0f-11eb-8539-a2901a12adee.png)

I created [this](https://codesandbox.io/s/cocky-morning-cvbbf?file=/src/store/ItemList.ts) sandbox

**Reply by danielkcz** [reply]
Easiest is `as any`, if you want strict, then `as IObservableArray<Item>`

**Reply by edshav** [reply]
Thank you, guys. Now it looks better:
```
import { IObservableArray, makeAutoObservable } from "mobx";
import { Item } from "./ItemModel";

export class ItemList {
  public items: IObservableArray<Item>;

  constructor(items: Item[] = []) {
    this.items = items as IObservableArray<Item>;
    makeAutoObservable(this);
  }

  addItem = (value: string) => {
    this.items.push(new Item(value, this));
  };

  removeItem = (itemToRemove: Item) => {
    this.items.remove(itemToRemove);
  };
}

```

**Reply by edshav** [reply]
And now we can use `this.items` without additional annotation in every method of the class
